pkg/backup/backuper.go

```diff
@@ -0,0 +1,354 @@
+package backup
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/mattn/go-sqlite3"
+
+	"github.com/pkg/errors"
+)
+
+// BackupFilenamePrefix is the prefix used in every backup file.
+const BackupFilenamePrefix = "tbl_backup"
+
+// Backuper is the process that executes the backup process.
+type Backuper struct {
+	sourcePath, dir string
+	source          DB
+	backup          DB
+	config          *Config
+
+	fileCreator func(string, time.Time) (string, error)
+}
+
+// NewBackuper creates a new backuper responsible for making backups of a SQLite database.
+func NewBackuper(sourcePath string, backupDir string, opts ...Option) (*Backuper, error) {
+	config := DefaultConfig()
+	for _, o := range opts {
+		if err := o(config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := os.MkdirAll(backupDir, 0o755); err != nil {
+		return nil, errors.Errorf("os mkdir all: %s", err)
+	}
+
+	b := &Backuper{
+		sourcePath:  sourcePath,
+		dir:         backupDir,
+		config:      config,
+		fileCreator: createBackupFile,
+	}
+
+	return b, nil
+}
+
+// Backup creates a backup to a file in disk.
+// Multiple serial calls to Backup can be perfomed. This can be used to perform retries in case of errors.
+func (b *Backuper) Backup(ctx context.Context) (_ BackupResult, err error) {
+	defer func() {
+		if err != nil {
+			_ = os.Remove(b.backup.Path())
+		}
+	}()
+
+	timestamp, err := b.init()
```

Comment:
`init` is now private and used inside Backup. This means that the backup file is created at the same time the process will be execute. Much cleaner API for clients.